From 9081f69b1ec7f436dc927feb625ead70b2ab7353 Mon Sep 17 00:00:00 2001
From: Evan Wang <xswang@marvell.com>
Date: Wed, 1 Mar 2017 19:09:05 +0800
Subject: [PATCH 0866/1345] ahci: mvebu: update sata suspend and resume
 routines in PM mode

commit  11032e845df5e5bbfa35e32e90e3a7610c0e42b2 from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

- The patch updates suspend routine that disable PHY
  gating clock, and disable regulators if there are.
- The patch updates resume routine to enable regulators,
  clock and PHY if there are, besides enable resources,
  it also adds SATA MAC config for Armada 3700 to power
  up the SATA MAC.

Change-Id: I35467f99cecdc7205c2770edbca8529006c97aca
Signed-off-by: Evan Wang <xswang@marvell.com>
Signed-off-by: Victor Gu <xigu@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/37139
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/ata/ahci_mvebu.c |   38 +++++++++++++++++++++++++++++++++-----
 1 file changed, 33 insertions(+), 5 deletions(-)

diff --git a/drivers/ata/ahci_mvebu.c b/drivers/ata/ahci_mvebu.c
index 56b6af4..fee5831 100644
--- a/drivers/ata/ahci_mvebu.c
+++ b/drivers/ata/ahci_mvebu.c
@@ -221,7 +221,18 @@ static int ahci_mvebu_pll_power_up(struct platform_device *pdev,
 #ifdef CONFIG_PM_SLEEP
 static int ahci_mvebu_suspend(struct platform_device *pdev, pm_message_t state)
 {
-	return ahci_platform_suspend_host(&pdev->dev);
+	int err;
+	struct ata_host *host = platform_get_drvdata(pdev);
+	struct ahci_host_priv *hpriv = host->private_data;
+
+	err = ahci_platform_suspend_host(&pdev->dev);
+	if (err)
+		return err;
+
+	/* AHCI resources, such as PHY, clock, etc. should be disabled */
+	ahci_platform_disable_resources(hpriv);
+
+	return 0;
 }
 
 static int ahci_mvebu_resume(struct platform_device *pdev)
@@ -229,12 +240,29 @@ static int ahci_mvebu_resume(struct platform_device *pdev)
 	struct ata_host *host = platform_get_drvdata(pdev);
 	struct ahci_host_priv *hpriv = host->private_data;
 	const struct mbus_dram_target_info *dram;
+	int err;
 
-	dram = mv_mbus_dram_info();
-	if (dram)
-		ahci_mvebu_mbus_config(hpriv, dram);
+	/* AHCI resources, such as PHY, clock, etc. should be enabled first */
+	err = ahci_platform_enable_resources(hpriv);
+	if (err)
+		return err;
+
+	if (of_device_is_compatible(pdev->dev.of_node,
+				    "marvell,armada-380-ahci")) {
+		dram = mv_mbus_dram_info();
+		if (dram)
+			ahci_mvebu_mbus_config(hpriv, dram);
 
-	ahci_mvebu_regret_option(hpriv);
+		ahci_mvebu_regret_option(hpriv);
+	}
+
+	if (of_device_is_compatible(pdev->dev.of_node,
+				    "marvell,armada-cp110-ahci"))
+		ahci_mvebu_pll_power_up(pdev, hpriv, 1);
+
+	if (of_device_is_compatible(pdev->dev.of_node,
+				    "marvell,armada-3700-ahci"))
+		ahci_mvebu_pll_power_up(pdev, hpriv, 0);
 
 	return ahci_platform_resume_host(&pdev->dev);
 }
-- 
1.7.9.5

